packages/vaex-core/vaex/expression.py

```diff
@@ -196,6 +197,148 @@ def __init__(self, expression):
         self.expression = expression
 
 
+class StructOperations(collections.abc.Mapping):
+    """Struct Array operations.
+
+    Usually accessed using e.g. `df.name.struct.get('field1')`
+
+    """
+
+    def __init__(self, expression):
+        self.expression = expression
+        self._array = self.expression.values
+
+    def __iter__(self):
+        for name in self.keys():
+            yield name
+
+    def __getitem__(self, key):
+        """Return struct field by either field name (string) or index position (index).
+
+        In case of ambiguous field names, a `LookupError` is raised.
+
+        """
+
+        self._assert_struct_dtype()
+        return self.get(key)
+
+    def __len__(self):
+        """Return the number of struct fields contained in struct array.
+
+        """
+
+        self._assert_struct_dtype()
+        return len(self._array.type)
+
+    def keys(self):
+        """Return all field names contained in struct array.
+
+        :returns: list of field names.
+
+        Example:
+
+        >>> import vaex
+        >>> import pyarrow as pa
+        >>> array = pa.StructArray.from_arrays(arrays=[[1,2], ["a", "b"]], names=["col1", "col2"])
+        >>> df = vaex.from_arrays(array=array)
+        >>> df
+        # 	array
+        0	{'col1': 1, 'col2': 'a'}
+        1	{'col1': 2, 'col2': 'b'}
+
+        >>> df.array.struct.keys()
+        ["col1", "col2"]
+
+        """
+
+        self._assert_struct_dtype()
+        return [field.name for field in self._array.type]
+
+    def values(self):
+        """Return all field types contained in struct array.
+
+        :returns: list of field types.
+
+        Example:
+
+        >>> import vaex
+        >>> import pyarrow as pa
+        >>> array = pa.StructArray.from_arrays(arrays=[[1,2], ["a", "b"]], names=["col1", "col2"])
+        >>> df = vaex.from_arrays(array=array)
+        >>> df
+        # 	array
+        0	{'col1': 1, 'col2': 'a'}
+        1	{'col1': 2, 'col2': 'b'}
+
+        >>> df.array.struct.values()
+        [int64, string]
+
+        """
+
+        self._assert_struct_dtype()
+        dtypes = (field.type for field in self._array.type)
+        vaex_dtypes = [DataType(x) for x in dtypes]
+
+        return vaex_dtypes
```

Comment:
I expected (since it's a mapping)
```suggestion
        return [self[i] for i in range(len(self))]
```

Reply:
Good point - I guess we've misunderstood us here. From my perspective, the main purpose of the `StructOperations` object was to represent meta data (field names & types) and to provide a namespace for all struct relevant functions. The mapping I was concerned about was between field names and field types.

If I want to get the actual values of the struct, I would rather use `df.x` (returning the vaex expression) or `df.x.values` (returning the pyarrow struct) instead of `df.x.struct.values()` (which would return a python list of pyarrow arrays in your version). However, I totally agree that the name `values` is misleading when in fact dtypes are returned. It depends on the kind of mapping we're thinking of (*field names -> types* vs. *field names -> arrays*). Both implementations are somewhat not satisfying really. Your proposal adds yet another way to access the data adding a 3rd return type while my approach is somewhat misleading regarding the naming convention and what a user might expect.

What about having `df.x.struct.values()` return `df.x.values`? This would be consistent in my opinion even though it makes `df.x.struct.values()` completely redundant. Hence, there is one way to return the actual vaex expression via `df.x` and two consistent ways to return the underlying data via `df.x.values` or `df.x.struct.values()`.

Reply:
The is exactly what I was referring to with my earlier comment about redundancy. 
I would tend to agree that `df.x.struct.values()` being the same as `df.x.values` is redundant, and does not really add anything in my opinion. 

@mansenfranzen if you like/need to keep the functionality of the curret `.struct.values()` that's fine, but perhaps  with a different name than `values` as that seems to be "reserved" for functionality of the type `df.x.values`. 

As for returning the data, i would cast my vote to just having `df.x.values` (which would return arrow thing). Dont know if the output of that would support something like `.tolist()` that would convert the arrow structure to list.. if there is, i say we are done. Otherwise maybe we can add that on our end?

Reply:
> I would tend to agree that `df.x.struct.values()` being the same as `df.x.values` is redundant, and does not really add anything in my opinion.

This is not the case right.

df.x.struct has a mapping interface, which should behave like a dict, which means that if `df.x.struct['col1']` gives an expression, `df.x.struct.values()` should give a list of expressions right?

Reply:
Yeah I agree with @maartenbreddels regarding `df.x.struct.values()`

But, i also do think that `df.x.values` should always exist, as there should be a uniform way to extract the raw data for any kind of column..

Reply:
Yeah, you've convinced me ;-). Honestly, my initial perspective of having a mapping for field names to field types for `StructOperations` seems rather inappropriate to me now. 

> if you like/need to keep the functionality of the curret .struct.values() that's fine, but perhaps with a different name than values as that seems to be "reserved" for functionality of the type df.x.values.

@JovanVeljanoski No need to keep that functionality. 

> df.x.struct has a mapping interface, which should behave like a dict, which means that if df.x.struct['col1'] gives an expression, df.x.struct.values() should give a list of expressions right?

@maartenbreddels You're right. I'm going to change it accordingly (probably tomorrow).
> Also, please don't confuse df.x.values with this, which should never be there :) (my personal opinion.. :))

The `df.x.values` seems to have some controversial opinions here ;-). Is there an issue or PR discussing it in more detail to follow along?

Reply:
> but perhaps with a different name than values as that seems to be "reserved" for functionality of the type df.x.values.

@JovanVeljanoski 
Just be be sure, `df.x.struct.values()` should implement what the mapping interface dictates (a list/sequence of Expressions), it has zero relation to `df.x.values` right

Reply:
Yeah, absolutely, we agree there. 
I was just saying that `df.x.values` has its place.